### ISSUE TITLE
Ability to group jobs?

### ISSUE THREAD

**Issue by jakubka** [OptimalBits/bull#1760]
I am wondering if bull would work for me.

In our software user has an option to modify multiple (up to 100k) entities at once. Modifying one entity means making one HTTP call, but that's an implementation detail. We need to have a queueing system for managing those jobs as they can take quite a long time.

`bull` looks super promising, but I didn't find an option to group jobs to achieve two things:
- jobs from different groups would be processed in parallel - so that one customer doesn't block other.. they just slow each other down
- know how many jobs from each group are done/failed/pending in order to be able to show the progress bar

In my head I am imagining creating a group of jobs every time user requests starts that bulk action and then periodically checking progress.

Is something like that possible with bull? 🤔

**Comment by EthanOrange** [comment]
Look like is cluster mode?

**Comment by sundowndev** [comment]
> Modifying one entity can be simplified to making an http call. We need to have a queueing system for managing those jobs as they can take quite a long time.

Why would you wait for the request to finish then ? The concept of background job is... to run tasks in the background. I'd probably use HTTP to create the task, then refresh items periodically. Using a worker to keep making things sync is non-sense to me.

**Comment by jakubka** [comment]
> Look like is cluster mode?

Yeah we'd run the queues in cluster mode to increase throughput, but still I think the feature to group jobs is not there.

**Comment by jakubka** [comment]
> Why would you wait for the request to finish then ? The concept of background job is... to run tasks in the background.

I didn't not phrase this well. What I wanted to say is that one job would be making one HTTP call. But that doesn't matter much as bull is generic. We can do whatever we want inside the job.

> I'd probably use HTTP to create the task, then refresh items periodically.

This is exactly how we plan to do it.

**Comment by sundowndev** [comment]
What you can also do is just create a queue to delete N items. The queue takes an array of IDs, your client make an HTTP call to your backend with N IDs as data, backend creates one task in this queue.

To follow progress, you can create a route on backend that'll retrieve the job's UUID inside Redis and check for the progress. But I think this feature is pointless.

**Comment by jakubka** [comment]
N can be 100k. Would bull handle it well if we had a task to process 100k items which can take tens of minutes? Hmm it doesn't sound like it was built for this. Every item should be one job/task in the queue I'd say.

**Comment by jakubka** [comment]
What about dynamically creating a queue for every bulk action? Is that something which bull would handle well? We'd most likely need to store the list of queues somewhere else, but that's easily doable.

**Comment by sundowndev** [comment]
Yes you're right, 100k is a lot, forget this solution then.

>Is that something which bull would handle well?

I don't know but I think Bull is based on Redis pub/sub. You can do your own benchmark, see this discussion about pub/sub performances: https://groups.google.com/forum/#!topic/redis-db/R09u__3Jzfk

**Comment by manast** [comment]
@jakubka 

| jobs from different groups would be processed in parallel - so that one customer doesn't block other.. they just slow each other down

Although not exactly the same, we are working in a new function that allows you to rate limit based on a job data key (https://github.com/OptimalBits/bull/pull/1753). This means you could put a limit on how many jobs a given "group" may be able to process per second. 

| know how many jobs from each group are done/failed/pending in order to be able to show the progress bar

I think this feature would match another feature request we had have for some time that allows you to define parent-child relationships.

**Comment by jakubka** [comment]
Nice 👍 I started watching #1753. Parent-child relationship sounds like they might do it. Is there any page with more details about this feature so I can check if it would fit?

**Comment by aramix** [comment]
@jakubka we are using bull for processing a queue with a couple of millions records and it seems to be able to handle them ok enough.

**Comment by DimosthenisK** [comment]
@jakubka any updates on this? Did you manage to do it somehow? I practically have the same use case and I am completely lost right now.

**Comment by jakubka** [comment]
@DimosthenisK We didn't go any further with this. We still have the need, but postponed the implementation in favour of other stuff and will get back to this later. Sorry!

**Comment by orenKingK** [comment]
Is there any solution to group bulk jobs in one queue?
